src/ert/scheduler/scheduler.py

```diff
@@ -181,32 +181,69 @@ async def execute(
         # cancel jobs from another thread
         self._loop = asyncio.get_running_loop()
-        async with background_tasks() as cancel_when_execute_is_done:
-            cancel_when_execute_is_done(self._publisher())
-            cancel_when_execute_is_done(self._process_event_queue())
-            cancel_when_execute_is_done(self.driver.poll())
-            if min_required_realizations > 0:
-                cancel_when_execute_is_done(
+        scheduling_tasks = [
+            asyncio.create_task(self._publisher()),
+            asyncio.create_task(self._process_event_queue()),
+            asyncio.create_task(self.driver.poll()),
+        ]
+
+        if min_required_realizations > 0:
+            scheduling_tasks.append(
+                asyncio.create_task(
                     self._stop_long_running_jobs(min_required_realizations)
                 )
-                cancel_when_execute_is_done(self._update_avg_job_runtime())
-
-            start = asyncio.Event()
-            sem = asyncio.BoundedSemaphore(self._max_running or len(self._jobs))
-            for iens, job in self._jobs.items():
-                self._tasks[iens] = asyncio.create_task(
-                    job(start, sem, self._max_submit)
-                )
-
-            start.set()
-            results = await asyncio.gather(
-                *self._tasks.values(), return_exceptions=True
             )
-            for result in results:
-                if isinstance(result, Exception):
-                    logger.error(result)
-
-            await self.driver.finish()
+            scheduling_tasks.append(asyncio.create_task(self._update_avg_job_runtime()))
+
+        start = asyncio.Event()
+        sem = asyncio.BoundedSemaphore(self._max_running or len(self._jobs))
+        for iens, job in self._jobs.items():
+            self._tasks[iens] = asyncio.create_task(job(start, sem, self._max_submit))
+        start.set()
+
+        async def gather_realization_jobs() -> list[BaseException | None]:
+            """This makes sure that all the tasks are completed, where afterwards
+            we cancel scheduling_tasks.It returns list of task exceptions or None.
+            """
+            try:
+                return await asyncio.gather(
+                    *self._tasks.values(), return_exceptions=True
+                )
+            finally:
+                for scheduling_task in scheduling_tasks:
+                    scheduling_task.cancel()
+
+        job_results: Optional[list[BaseException | None]] = None
+        try:
+            # we gather collectively results of realization tasks and scheduling_tasks
+            # If Exception/CancellationError were to happen we know it's coming
+            # from scheduling_tasks only and thus we need to cancel realization tasks.
+            # Otherwise potential exceptions from realization tasks
+            # are stored in job_results and handled in else branch
```

Comment:
"we gather collectively results..."?

```suggestion
            # `gather_realization_jobs` does not raise; rather, each job task's result
            # is collected into the returning list. If an exception occurs, it must
            # necessarily come from the scheduling tasks. Exceptions from the
            # job tasks are handled in the `else` branch after the evaluation
            # has stopped
```